### ISSUE TITLE
Call Deflection Issue

### ISSUE THREAD

**Issue by robyevolution** [mammuth/ha-fritzbox-tools#73]
Hi I can't activate call forwarding, I have the component updated to the latest version and I get this error from the log:

2020-02-18 08:59:08 ERROR (SyncWorker_5) [custom_components.fritzbox_tools.switch] Call Deflection switches could not be enabled.
Traceback (most recent call last):
  File "/home/homeassistant/.homeassistant/custom_components/fritzbox_tools/switch.py", line 43, in _create_deflection_switches
    fritzbox_tools, item, dict_of_deflection
  File "/home/homeassistant/.homeassistant/custom_components/fritzbox_tools/switch.py", line 303, in __init__
    self.id = self.dict_of_deflection["DeflectionId"]
TypeError: list indices must be integers or slices, not str

**Comment by AaronDavidSchneider** [comment]
Hi, thanks for your report! I cant see why you get a list instead of a dict.
I will have a look into it. Out of curiousity: Do you have multiple call forwarding settings in your fritzbox?

**Comment by robyevolution** [comment]
i have only 1 deflection

**Comment by AaronDavidSchneider** [comment]
which version of homeassistant do you use? The new versions of the component require home assistant 0.105. 
Which Fritzbox do you use?

I can't reproduce the error on my setup.

Do you have some python knowledge? If so, could you please try the following:

in your terminal:
```
pip install fritzconnection==1.2.0
```

with python:
```py
import xmltodict
from fritzconnection import FritzConnection

fc = FritzConnection(
            address="...", port="...", user="...", password="...", timeout=30.0
        )
xmltodict.parse(
                    fc.call_action("X_AVM-DE_OnTel:1", "GetDeflections")["NewDeflectionList"]
                )["List"].items()
```

In the above you need to replace "...", by username, password and ip of your fritzbox

**Comment by robyevolution** [comment]
I have the latest version of home assistant but the error also appeared in previous versions, the component fritzconnection == 1.2.0 was already updated.
I have a 7590 fritzbox with the latest firmware (no beta) installed.

**Comment by AaronDavidSchneider** [comment]
Thanks for your answer. Could you try to run the above script? And post the response?
Note: If you dont use ipython, you need to adapt the last line to:
```
print(xmltodict.parse(
                    fc.call_action("X_AVM-DE_OnTel:1", "GetDeflections")["NewDeflectionList"]
                )["List"].items())
```

**Comment by AaronDavidSchneider** [comment]
The response should look like this:

```py
odict_items([('Item', OrderedDict([('DeflectionId', '0'), ('Enable', '0'), ('Type', 'fromAll'), ('Number', None), ('DeflectionToNumber', '01783161990'), ('Mode', 'eImmediately'), ('Outgoing', None), ('PhonebookID', None)]))])
```

**Comment by robyevolution** [comment]
how do I start the script?

**Comment by AaronDavidSchneider** [comment]
if you have python installed on your computer (I use anaconda: https://www.anaconda.com/distribution/):

go to your terminal:
```
pip install fritzconnection==1.2.0
pip install xmltodict
```

open ipython: copy code from above in there and hit enter

**Comment by robyevolution** [comment]
I'm not a python expert, I use home assistant on a raspberry pi 4 with raspian buster, if you show me all the steps to do I try

**Comment by AaronDavidSchneider** [comment]
sure, no problem! I suppose you use a venv then (as explained here: https://www.home-assistant.io/docs/installation/virtualenv/)? Or do you use a different setup?

you would then simply need to edit the script below with your credentials (port should be 49000) and save it as a .py file (like test_deflections.py)

```py
import xmltodict
from fritzconnection import FritzConnection

fc = FritzConnection(
            address="...", port=49000, user="...", password="...", timeout=30.0
        )
print(xmltodict.parse(
                    fc.call_action("X_AVM-DE_OnTel:1", "GetDeflections")["NewDeflectionList"]
                )["List"].items())
```

Activate your virtual invironment: (go to the folder where homeassistant is located)
```
cd homeassistant
source bin/activate
```

and hit the script: `python test_deflections.py`
If it says that a package is missing install it with the above instruction.

I hope this helps :D

**Comment by robyevolution** [comment]
I followed the instructions, here's what comes out:

(homeassistant) homeassistant@hassbian:~/.homeassistant/python_scripts $ python deflection.py
odict_items([('Item', [OrderedDict([('DeflectionId', '0'), ('Enable', '1'), ('Type', 'fromNumber'), ('Number', '00'), ('DeflectionToNumber',                                                   None), ('Mode', 'eNoSignal'), ('Outgoing', None), ('PhonebookID', None)]), OrderedDict([('DeflectionId', '1'), ('Enable', '1'), ('Type', 'fro                                                  mNumber'), ('Number', '04'), ('DeflectionToNumber', None), ('Mode', 'eNoSignal'), ('Outgoing', None), ('PhonebookID', None)]), OrderedDict([(                                                  'DeflectionId', '2'), ('Enable', '1'), ('Type', 'fromNumber'), ('Number', '05'), ('DeflectionToNumber', None), ('Mode', 'eNoSignal'), ('Outgo                                                  ing', None), ('PhonebookID', None)]), OrderedDict([('DeflectionId', '3'), ('Enable', '1'), ('Type', 'fromNumber'), ('Number', '3906'), ('Defl                                                  ectionToNumber', None), ('Mode', 'eNoSignal'), ('Outgoing', None), ('PhonebookID', None)]), OrderedDict([('DeflectionId', '4'), ('Enable', '1                                                  '), ('Type', 'fromNumber'), ('Number', '055'), ('DeflectionToNumber', None), ('Mode', 'eNoSignal'), ('Outgoing', None), ('PhonebookID', None)                                                  ]), OrderedDict([('DeflectionId', '5'), ('Enable', '1'), ('Type', 'fromNumber'), ('Number', '09'), ('DeflectionToNumber', None), ('Mode', 'eN                                                  oSignal'), ('Outgoing', None), ('PhonebookID', None)]), OrderedDict([('DeflectionId', '6'), ('Enable', '1'), ('Type', 'fromNumber'), ('Number                                                  ', '0381'), ('DeflectionToNumber', None), ('Mode', 'eNoSignal'), ('Outgoing', None), ('PhonebookID', None)]), OrderedDict([('DeflectionId', '                                                  7'), ('Enable', '0'), ('Type', 'fromAll'), ('Number', None), ('DeflectionToNumber', '600'), ('Mode', 'eImmediately'), ('Outgoing', None), ('P                                                  honebookID', None)])])])

**Comment by AaronDavidSchneider** [comment]
Thankyou! This helps!
As you can see in your response, you actually have 8 deflections instead of one. I didn't test this case. But I now understand the problem and know how to fix it. I will create a Pull request and fix the problem within the next hours.
Thanks again for reporting the bug.

**Comment by AaronDavidSchneider** [comment]
Could you please test the changes in my PR #74 @robyevolution?
To do so you simply need to replace your switch.py file with this one: https://github.com/AaronDavidSchneider/ha-fritzbox-tools/blob/fix_deflections/custom_components/fritzbox_tools/switch.py

**Comment by robyevolution** [comment]
(homeassistant) homeassistant@hassbian:~/.homeassistant/python_scripts $ python switch.py
Traceback (most recent call last):
  File "switch.py", line 15, in <module>
    from . import DOMAIN, DATA_FRITZ_TOOLS_INSTANCE
ImportError: cannot import name 'DOMAIN' from '__main__' (switch.py)

**Comment by AaronDavidSchneider** [comment]
you don't need to execute switch.py manually. You just need to replace it in your custom_components/fritzbox_tools folder and restart home assistant :D

**Comment by robyevolution** [comment]
Now it works perfectly!

**Comment by AaronDavidSchneider** [comment]
Glad it works now. Thankyou very much for your help!

**Comment by mammuth** [comment]
Closed by https://github.com/mammuth/ha-fritzbox-tools/pull/74 which is released as version `1.8.1` on HACS.

Thanks for bringing up the issue and helping @AaronDavidSchneider to debug it @robyevolution a! 👍
